src/hutchinson.jl

```diff
@@ -1,59 +1,135 @@
 # Hutchinson Trace Estimator for Inverse of Matrix
 # Based on Hutchinson's Stochastic Estimation Proof
 # Values from Extrapolation Methods for Estimating Trace of the Matrix Inverse by P. Fika
+# (This method works for SPD with low condition number)
 
-export hutch
+export hutch, hutch!, HutchWorkspace
 
 using LinearAlgebra
 
+struct HutchWorkspace
+    A::AbstractArray{<:Any, 2}
+    randfunc::Function
+    x::AbstractArray{<:Any, 1}
+    y::AbstractArray{<:Any, 1}
+    N::Int64
+    skipverify::Bool
+end
+
+"""
+    HutchWorkspace(A; N = 30, skipverify = false)
+
+# Arguments
+ - `A` : Symmetric Hermitian Matrix with low condition number
+ - `N` : Number of iterations (Default: 30)
+ - `skipverify` : If false, it will check isposdef(A) (Default: false)
+"""
+function HutchWorkspace(A; N = 30, skipverify = false)
+    randfunc(n) = rand(-1.0:2.0:1.0, n)
+    x = randfunc(size(A)[1])
+    y = similar(x)
+    return HutchWorkspace(A, randfunc, x, y, N, skipverify)
+end
+
+"""
+    HutchWorkspace(A, randfunc::Function; N = 30, skipverify = false)
+
+# Arguments
+ - `A` : Symmetric Hermitian Matrix with low condition number
+ - `randfunc` : Function to generate random values for x
+                Distributed uniformly 
+                (Base: rand(-1.0:2.0:1.0, size(A)[1]))
+                Example: f(n) = rand(-1.0:2.0:1.0, n)
+ - `N` : Number of iterations (Default: 30)
+ - `skipverify` : If false, it will check isposdef(A) (Default: false)
+"""
+function HutchWorkspace(A, randfunc::Function; N = 30, skipverify = false)
+    x = randfunc(size(A)[1])
+    y = similar(x)
+    return HutchWorkspace(A, randfunc, x, y, N, skipverify)
+end
+
 # Calculating the moments and extrapolating c-1 
-function ev(A)
-    # Create a vector with +-1 values with 0.5 probability 
-    x = rand(-1:2:1, size(A)[1])
-    
+function ev(w::HutchWorkspace)
+    # Create new random values for x
+    copyto!(w.x, w.randfunc(size(w.A)[1]))
```

Comment:
This line will allocate for the output of `w.randfunc`. If `Base.rand` is passed, we can do:
```julia
T = eltype(w.x)
o = one(T)
t = 2*o
w.x .= rand.(Ref(-o:t:o))
```
Try to understand the last line. `Ref` "protects" the range when broadcasting. Note that other `rand` functions for special array types support broadcasting too, so this should work for `CuArrays.rand` too.